src/Streamly/Internal/Data/Stream/Async/Channel/Dispatcher.hs

```diff
@@ -62,65 +62,102 @@ pushWorker yieldMax sv = do
                     , workerYieldCount = cntRef
                     , workerLatencyStart = lat
                     }
-    doFork (workLoop sv winfo) (svarMrun sv) exception
-        >>= addThread (workerThreads sv)
+    -- In case of lazy dispatch we dispatch workers only from the consumer
+    -- thread. In that case it is ok to use addThread here as it is guaranteed
+    -- that the thread will be added to the workerSet before the thread STOP
+    -- event is processed, because we do both of these actions in the same
+    -- consumer thread. However, in case of eager dispatch we may dispatch
+    -- workers from workers, in which case the thread Stop even may get
+    -- processed before the addThread occurs, so in that case we have to use
+    -- modifyThread which performs a toggle rather than adding or deleting.
+    --
+    -- XXX We can use addThread or modThread based on eager flag.
+    doFork (workLoop sv winfo) (svarMrun sv) exception >>= modThread
 
     where
 
+    modThread = modifyThread (workerThreads sv) (outputDoorBell sv)
     exception = handleChildException (outputQueue sv) (outputDoorBell sv)
 
--- Returns:
--- True: can dispatch more
--- False: cannot dispatch any more
-dispatchWorker :: (MonadIO m, MonadBaseControl IO m) => Count -> Channel m a -> m Bool
-dispatchWorker yieldCount sv = do
+{-# INLINE getEffectiveWorkerLimit #-}
+getEffectiveWorkerLimit :: MonadIO m => Channel m a -> m Limit
+getEffectiveWorkerLimit sv = do
     let workerLimit = maxWorkerLimit sv
+    case remainingWork sv of
+        Nothing -> return workerLimit
+        Just ref -> do
+            n <- liftIO $ readIORef ref
+            case yieldRateInfo sv of
+                Just _ -> return workerLimit
+                Nothing ->
+                    return $
+                        case workerLimit of
+                            Unlimited -> Limited (fromIntegral n)
+                            Limited lim -> Limited $ min lim (fromIntegral n)
+
+{-# INLINE checkMaxThreads #-}
+checkMaxThreads :: MonadIO m => Int -> Channel m a -> m Bool
+checkMaxThreads active sv = do
+    -- Note that we may deadlock if the previous workers (tasks in the
+    -- stream) wait/depend on the future workers (tasks in the stream)
+    -- executing. In that case we should either configure the maxWorker
+    -- count to higher or use parallel style instead of ahead or async
+    -- style.
```

Comment:
Can you write about what `Config` options to use, instead of the word `parallel`.

Reply:
This is an old comment, still kind of valid, because we do have parallel, ahead, async. We can change it in future.